### ISSUE TITLE
kube-proxy/iptables, client continuous to send packets, udp server pod restart cause UDP conntrack entry error

### ISSUE THREAD

**Issue by lyyao09** [kubernetes/kubernetes#102618]
#### What happened:
Devices outside the cluster frequently send udp packets to the pod, which causes the source ip in the contrack entry to be set to the node ip after the pod restarts and cannot be aged. The problem still exists on the k8s 1.21.0.

#### What you expected to happen:
The source ip in the contrack entry records podip instead of node ip.
#### How to reproduce it (as minimally and precisely as possible):
1. Install k8s cluster;
```
[root@node]# kubectl get node -owide
NAME    STATUS   ROLES   VERSION    INTERNAL-IP     KERNEL-VERSION               CONTAINER-RUNTIME
node    Ready     master   v1.15.12    10.10.212.164    3.10.0-957.27.2.el7.x86_64   docker://18.9.6
```
2. Deploy a udp service exposed through nodeport, eg；
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: dao
spec:
  replicas: 1
  selector:
    matchLabels:
      app: dao
  template:
    metadata:
      labels:
        app: dao
    spec:
      containers:
      - image: samwelkey24/dao-2048:1.0
        name: dao
---
apiVersion: v1
kind: Service
metadata:
  name: dao
  labels:
    app: dao
spec:
  type: NodePort
  ports:
  - port: 80
    targetPort: 80
    name: tcp
  - port: 8080
    targetPort: 8080
    nodePort: 30030
    name: udp
    protocol: UDP
  selector:
    app: dao
```
3. Use the nc command to simulate the client frequently sending udp packets outside the cluster, eg；
```
[root@node]# while true; do echo "test" | nc -4u  10.10.212.164 30030 -p 9999;done
```
4. Delete pod, eg；
```
[root@node]#  kubectl get pod -n allkinds -owide
NAME                                 READY   STATUS    RESTARTS   AGE   IP                NODE       
dao-5f7669bc69-kkfk5   1/1     Running   0          18m   177.177.241.185   node

[root@node]# kubectl delete pod dao-5f7669bc69-kkfk5
```
5. View the contrack entry of the specified nodeport port, eg 30030, eg；
```
normal entry：
[root@node]# cat /pro/net/nf_contrack |grep 30030
ipv4     2 udp      17 29 src=10.10.212.167 dst=10.10.212.164 sport=9999 dport=30030 [UNREPLIED] src=177.177.241.185 dst=10.10.212.164 sport=8080 dport=9999 mark=0 zone=0 use=2

error entry：
[root@node]# cat /pro/net/nf_contrack |grep 30030
ipv4     2 udp      17 29 src=10.10.212.167 dst=10.10.212.164 sport=9999 dport=30030 [UNREPLIED] src=10.10.212.164 dst=10.10.212.167 sport=8080 dport=9999 mark=0 zone=0 use=2
```
6. After pod restarts, udp packets can no longer be received, eg；
```
[root@node]# tcpdump -n -i cali781669ea248 port 8080
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
^
```
#### Anything else we need to know?:
According to my positioning, when the pod is deleted, the redundant udp entries will also be deleted, but during the pod start, the reject rule in the filter table will be cleaned up, and then the dnat rule will be added. This sequence of operations will probabilistically cause contrack entry errors.

related source code:
https://github.com/kubernetes/kubernetes/blob/8eff96dd12b5bde0fdf2d5d165c9d3575bc63d99/pkg/proxy/iptables/proxier.go#L1613-L1616

By adjusting the operation sequence of the filter table and the nat table, the problem can be solved. I would like to know if there is any functional impact of adjusting this order? Or is there a better solution?

#### Environment:
- Kubernetes version (use `kubectl version`):  1.15.12 and 1.20.1
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):  CentOS 7.6
- Kernel (e.g. `uname -a`): 3.10.0-957.21.3.el7.x86_64
- Install tools: 
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by uablrek** [comment]
Good analyze :+1: 

There are some conntrack related issues right now. I think it is a good option to remove *all* entries with the NATed src set to a loadBalancerIP or clusterIP. Such an entry can't be right. For all protocols, see also https://github.com/kubernetes/kubernetes/issues/101607#issuecomment-853965545.

**Comment by uablrek** [comment]
I triage this as a bug along the other conntrack bugs.

/triage accepted

**Comment by aojea** [comment]
I think that I fixed this race recently https://github.com/kubernetes/kubernetes/pull/98305, in 1.21

**Comment by lyyao09** [comment]
> I think that I fixed this race recently #98305, in 1.21, when the service goes from 0 endpoints to more than 1 endpoint, conntrack entries are flushed

Great, it seems to be fixed, wait for me to verify and then feedback.

**Comment by lyyao09** [comment]
:+1:   occasionally there is an abnormal entry, and then it will be deleted.
```
ipv4     2 udp      17 29 src=10.10.212.167 dst=10.10.212.164 sport=9999 dport=30030 [UNREPLIED] src=10.10.212.164 dst=10.10.212.167 sport=8080 dport=9999 mark=0 zone=0 use=2
ipv4     2 udp      17 29 src=10.10.212.167 dst=10.10.212.164 sport=9999 dport=30030 [UNREPLIED] src=177.177.241.185 dst=10.10.212.164 sport=8080 dport=9999 mark=0 zone=0 use=2
ipv4     2 udp      17 29 src=10.10.212.167 dst=10.10.212.164 sport=9999 dport=30030 [UNREPLIED] src=177.177.241.185 dst=10.10.212.164 sport=8080 dport=9999 mark=0 zone=0 use=2
ipv4     2 udp      17 29 src=10.10.212.167 dst=10.10.212.164 sport=9999 dport=30030 [UNREPLIED] src=177.177.241.185 dst=10.10.212.164 sport=8080 dport=9999 mark=0 zone=0 use=2
```
